File: src/main/java/wraith/fabricaeexnihilo/api/registry/OreRegistry.java
package wraith.fabricaeexnihilo.api.registry;

import net.minecraft.client.util.ModelIdentifier;
import org.jetbrains.annotations.Nullable;
import wraith.fabricaeexnihilo.modules.ore.ChunkMaterial;
import wraith.fabricaeexnihilo.modules.ore.ChunkShape;
import wraith.fabricaeexnihilo.modules.ore.PieceShape;
import wraith.fabricaeexnihilo.modules.ore.OreProperties;
import wraith.fabricaeexnihilo.util.Color;

import java.util.List;

public interface OreRegistry extends Registry<OreProperties> {
    List<OreProperties> getAll();

    boolean register(OreProperties... properties);
    default boolean register(String material, Color color, PieceShape pieceShape, ChunkShape chunkShape, ChunkMaterial chunkMaterial) {
        var oreProperties =
                OreProperties.Builder(material)
                        .setColor(color)
                        .setPieceShape(pieceShape)
                        .setChunkShape(chunkShape)
                        .setChunkMaterial(chunkMaterial)
                .build();
        return register(oreProperties);
    }

    void registerPieceItems();
    void registerChunkItems();

    @Nullable
    OreProperties getPropertiesForModel(ModelIdentifier identifier);
}

File: src/main/kotlin/exnihilofabrico/registry/OreRegistry.kt
package exnihilofabrico.registry

import exnihilofabrico.ExNihiloFabrico
import exnihilofabrico.api.registry.IOreRegistry
import exnihilofabrico.modules.ore.OreChunkItem
import exnihilofabrico.modules.ore.OrePieceItem
import exnihilofabrico.modules.ore.OreProperties
import net.minecraft.client.util.ModelIdentifier
import net.minecraft.item.Item
import net.minecraft.util.registry.Registry

data class OreRegistry(val registry: MutableList<OreProperties> = mutableListOf()): IOreRegistry {

    val item_settings: Item.Settings = Item.Settings().group(ExNihiloFabrico.ITEM_GROUP).maxCount(64)

    override fun clear() = registry.clear()
    override fun getAll() = registry
    override fun register(vararg properties: OreProperties) = registry.addAll(properties)

    override fun registerPieceItems(itemRegistry: Registry<Item>) =
        registry.forEach { Registry.register(itemRegistry, it.getPieceID(), OrePieceItem(it, item_settings)) }

    override fun registerChunkItems(itemRegistry: Registry<Item>) =
        registry.forEach { Registry.register(itemRegistry, it.getChunkID(), OreChunkItem(it, item_settings)) }

    override fun getPropertiesForModel(identifier: ModelIdentifier): OreProperties? =
        registry.firstOrNull { it.getChunkID().path == identifier.path || it.getPieceID().path == identifier.path }
}